### PR TITLE
[BEAM-2076] Shade Runner Dependencies in the DirectRunner

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---
Move inner classes of the DirectRunner to reduce total API Surface.

Add API Surface Test for the Direct Runner.

### CODE REVIEW

File: runners/direct-java/pom.xml [hunk 1]
```diff
@@ -87,6 +87,51 @@
...
+                      <exclude>META-INF/*.RSA</exclude>
+                    </excludes>
+                  </filter>
+                </filters>
+                <relocations>
+                  <relocation>
+                    <pattern>org.apache.beam.runners</pattern>
+                    <!--suppress MavenModelInspection -->
+                    <shadedPattern>
+                      org.apache.beam.runners.direct.util
```
Comment (lukecwik) [inline runners/direct-java/pom.xml]: util?

File: runners/direct-java/pom.xml [hunk 2]
```diff
@@ -87,6 +87,51 @@
         </executions>
       </plugin>
 
+        <plugin>
+          <groupId>org.apache.maven.plugins</groupId>
+          <artifactId>maven-shade-plugin</artifactId>
+          <version>3.0.0</version>
```
Comment (dhalperi) [inline runners/direct-java/pom.xml]: no version numbers.

Comment (tgroh) [inline runners/direct-java/pom.xml]: rm'd

File: runners/direct-java/pom.xml (hunk 1 above)
Comment (tgroh) [inline runners/direct-java/pom.xml]: *.util -> *.repackaged.runners.util

**Comment by tgroh** [comment]
(Removing test and runtime dependencies)

[INFO] --- maven-dependency-plugin:3.0.0:tree (default-cli) @ beam-runners-direct-java ---
[INFO] org.apache.beam:beam-runners-direct-java:jar:0.7.0-SNAPSHOT
[INFO] +- org.apache.beam:beam-sdks-java-core:jar:0.7.0-SNAPSHOT:compile
...
[INFO] +- org.apache.beam:beam-runners-core-construction-java:jar:0.7.0-SNAPSHOT:compile
[INFO] |  \- org.apache.beam:beam-sdks-common-runner-api:jar:0.7.0-SNAPSHOT:compile
[INFO] +- org.apache.beam:beam-runners-core-java:jar:0.7.0-SNAPSHOT:compile
[INFO] +- com.google.guava:guava:jar:20.0:compile
...
[INFO] +- joda-time:joda-time:jar:2.4:compile
[INFO] +- com.google.code.findbugs:jsr305:jar:3.0.1:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.14:compile
[INFO] +- com.google.auto.service:auto-service:jar:1.0-rc2:compile (optional) 
[INFO] |  \- com.google.auto:auto-common:jar:0.3:compile (optional) 
[INFO] +- com.google.auto.value:auto-value:jar:1.4.1:provided
[INFO] +- org.hamcrest:hamcrest-all:jar:1.3:provided
[INFO] +- junit:junit:jar:4.12:provided
...

We shade and repackage guava, runners-core-construction-java and runners-core-java. It may also be appropriate to repackage the runner API. My belief is that all other dependencies are also dependencies of the Java SDK, and thus cannot be repackaged.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/** API surface verification for {@link org.apache.beam.runners.direct}. */
+@RunWith(JUnit4.class)
+public class DirectRunnerApiSurfaceTest {
+  @Test
+  public void testDirectRunnerApiSurface() throws Exception {
+    // The DirectRunner should expose no more than the Core SDK
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java]: looks like this comment is now out of date?

File: sdks/java/core/src/test/java/org/apache/beam/SdkCoreApiSurfaceTest.java [hunk 4]
```diff
@@ -30,27 +30,26 @@
 /** API surface verification for {@link org.apache.beam}. */
 @RunWith(JUnit4.class)
 public class SdkCoreApiSurfaceTest {
+  @SuppressWarnings("unchecked")
+  public static final Set<String> ALLOWED_PACKAGES =
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/SdkCoreApiSurfaceTest.java]: no longer needed in this PR?

**Comment by tgroh** [comment]
Whoops, ran on the top-level POM.

We don't have a direct dependency on proto; right now we get a transitive dependency via the core SDK and runners-core-construction (and probably runners-core). Those can (and probably should) be shaded independently

[INFO] --- maven-dependency-plugin:3.0.0:tree (default-cli) @ beam-runners-direct-java ---
[INFO] org.apache.beam:beam-runners-direct-java:jar:0.7.0-SNAPSHOT
[INFO] +- org.apache.beam:beam-sdks-java-core:jar:0.7.0-SNAPSHOT:compile
[INFO] |  +- com.google.http-client:google-http-client:jar:1.22.0:compile
[INFO] |  |  \- org.apache.httpcomponents:httpclient:jar:4.0.1:compile
[INFO] |  |     +- org.apache.httpcomponents:httpcore:jar:4.0.1:compile
[INFO] |  |     +- commons-logging:commons-logging:jar:1.1.1:compile
[INFO] |  |     \- commons-codec:commons-codec:jar:1.3:compile
[INFO] |  +- com.google.protobuf:protobuf-java:jar:3.2.0:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.8.8:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-annotations:jar:2.8.8:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.8.8:compile
[INFO] |  +- net.bytebuddy:byte-buddy:jar:1.6.8:compile
[INFO] |  +- org.apache.avro:avro:jar:1.8.1:compile
[INFO] |  |  +- org.codehaus.jackson:jackson-core-asl:jar:1.9.13:compile
[INFO] |  |  +- org.codehaus.jackson:jackson-mapper-asl:jar:1.9.13:compile
[INFO] |  |  +- com.thoughtworks.paranamer:paranamer:jar:2.7:compile
[INFO] |  |  \- org.tukaani:xz:jar:1.5:compile
[INFO] |  +- org.xerial.snappy:snappy-java:jar:1.1.4-M3:compile
[INFO] |  +- org.apache.commons:commons-compress:jar:1.9:compile
[INFO] |  \- org.apache.commons:commons-lang3:jar:3.5:compile
[INFO] +- com.google.guava:guava-testlib:jar:20.0:test
[INFO] |  \- com.google.guava:guava:jar:20.0:compile
[INFO] +- joda-time:joda-time:jar:2.4:compile
[INFO] +- com.google.code.findbugs:jsr305:jar:3.0.1:compile
[INFO] +- org.slf4j:slf4j-api:jar:1.7.14:compile
[INFO] +- com.google.auto.service:auto-service:jar:1.0-rc2:compile (optional) 
[INFO] |  \- com.google.auto:auto-common:jar:0.3:compile (optional)

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java (hunk 3 above)
Comment (tgroh) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java]: updated

File: sdks/java/core/src/test/java/org/apache/beam/SdkCoreApiSurfaceTest.java (hunk 4 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/SdkCoreApiSurfaceTest.java]: rm'd

**Comment by tgroh** [comment]
Protobuf is shaded.

**Comment by tgroh** [comment]
R: @dhalperi

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
The failure was in archetype generation, which appears unrelated.

**Comment by tgroh** [comment]
retest this please

**Comment by tgroh** [comment]
PTAL

I needed to include the Runner API protos to shade `runners-core`, as they aren't otherwise on the classpath, and then I'm shading them so the dependency doesn't leak.

File: runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+/** API surface verification for {@link org.apache.beam.runners.direct}. */
+@RunWith(JUnit4.class)
+public class DirectRunnerApiSurfaceTest {
+  @Test
+  public void testDirectRunnerApiSurface() throws Exception {
+    // The DirectRunner can expose the Core SDK, anything exposed by the Core SDK, and itself
+    @SuppressWarnings("unchecked")
+    final Set<String> allowed =
+        ImmutableSet.of("org.apache.beam.sdk", "org.apache.beam.runners.direct", "org.joda.time");
```
Comment (dhalperi) [inline runners/direct-java/src/test/java/org/apache/beam/runners/direct/DirectRunnerApiSurfaceTest.java]: For future improvement, would be nice to make this be just specific classes `Pipeline`, `PipelineOptions`, `PipelineRunner`, and maybe that's roughly it.

**Comment by dhalperi** [comment]
LGTM, please self-merge and CP to release once green.

**Pull request closed without merging**